### ISSUE TITLE
opentelekomcloud_waf_certificate_v1 cannot be changed if in use by opentelekomcloud_waf_domain_v1

### ISSUE THREAD

**Issue by DasLeo** [opentelekomcloud/terraform-provider-opentelekomcloud#1278]
# Terraform Version

- 1.0.0

# Affected Resource(s)

- opentelekomcloud_waf_certificate_v1
- opentelekomcloud_waf_domain_v1

# Terraform Configuration Files

```hcl
resource "opentelekomcloud_waf_policy_v1" "policy_1" {
  name = "policy_1"
  options {
    webattack       = true
    common          = true
    crawler         = true
    crawler_engine  = true
    crawler_scanner = true
    crawler_script  = true
    crawler_other   = true
    webshell        = true
    cc              = true
    custom          = true
    whiteblackip    = true
    privacy         = true
    ignore          = true
    antitamper      = true
  }
  full_detection = false
  action {
    category = "log"
  }
}

resource "opentelekomcloud_waf_preciseprotection_rule_v1" "allowlist" {
  count     = length(var.waf_cidr_allowlist) > 0 ? 1 : 0
  policy_id = opentelekomcloud_waf_policy_v1.policy_1.id
  name      = "allowlist"
  dynamic "conditions" {
    for_each = var.waf_cidr_allowlist
    content {
      category = "ip"
      contents = [conditions.key]
      logic    = "equal"
    }
  }
  action_category = "block"
  priority        = 2
}

resource "opentelekomcloud_waf_domain_v1" "domain" {
  hostname = var.waf_domain
  server {
    client_protocol = "HTTPS"
    server_protocol = "HTTPS"
    address         = var.elb_public_ip
    port            = "443"
  }
  certificate_id = opentelekomcloud_waf_certificate_v1.certificate.id
  proxy          = false
  cipher         = "cipher_2"
  tls            = "TLS v1.2"
  policy_id      = opentelekomcloud_waf_policy_v1.policy_1.id
}
```

[...]

```hcl
resource "opentelekomcloud_waf_certificate_v1" "certificate" {
  name    = replace(var.waf_domain, ".", "-")
  content = sensitive(base64decode(var.waf_certificate.content))
  key     = sensitive(base64decode(var.waf_certificate.key))
}
```

[..]

```hcl
variable "tld" {
  type        = string
  description = "Tld"
}

variable "environment" {
  type        = string
  description = "environment"
}

variable "elb_public_ip" {
  type        = string
  description = "Public IP of ELB"
}

variable "waf_domain" {
  type        = string
  description = "Domain for WAF"
}

variable "waf_certificate" {
  type        = map(any)
  description = "Certificate and key for WAF"
  sensitive   = true
}

variable "waf_cidr_allowlist" {
  type        = map(string)
  description = "map of cidr to allow access"
}
```

[...]

```hcl
provider "opentelekomcloud" {
  access_key  = var.access_key
  secret_key  = var.secret_key
  domain_name = var.domain_name
  tenant_name = var.project_name
  auth_url    = var.auth_url
}

module "infrastructure" {
  source                     = "../../infrastructure"
  environment                = var.environment
  tld                        = var.tld
  availability_zones         = var.availability_zones
  multi_az                   = var.multi_az
  project_name               = var.project_name
  rds_disk_size              = var.rds_disk_size
  rds_flavor                 = var.rds.settings.flavor
  rds_availability_zone      = var.rds.settings.availability_zone
  rds_ha_replication_mode    = var.rds.settings.ha_replication_mode
  parametergroup_template    = var.parametergroup_template
  kms_key                    = var.kms_key
  nodepool_configuration     = var.nodepool_configuration
  nat_eip_publicip_type      = var.nat_eip_publicip_type
  modules_toggles            = merge(local.default_module_toggles, try(var.environment_modules_toggles, null))
  cbr_settings               = tomap(merge(local.default_cbr_settings, try(var.environment_cbr_settings, null)))
  nat_gateway_spec           = var.nat_gateway_spec
  vpc_additional_dns_servers = var.vpc_additional_dns_servers
  waf_custom_certificate     = var.waf_custom_certificate != null ? var.waf_custom_certificate : null
  waf_domain                 = var.waf_domain
  waf_cidr_allowlist         = var.waf_cidr_allowlist
}
```

[...]

```hcl
waf_custom_certificate = {
    "content": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx==",
    "key": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx=="
}
environment_modules_toggles = {
  waf = true
}
waf_domain = "foobar.example.com"
```

# Debug Output/Panic Output

[https://termbin.com/k6zh](https://termbin.com/k6zh)

# Steps to Reproduce
1. `terraform apply`
2. Adjust the `content` and `key` of the `certificate` resource
3. `terraform apply`

# Expected Behavior

The certificate object should be adjusted so that it reflects the new content.

# Actual Behavior

Changing the content of a `opentelekomcloud_waf_certificate_v1` results in an error, if it is in use by a WAF domain.
The way the provider works, makes it impossible to change SSL certificates, when they expire, because the `opentelekomcloud_waf_certificate_v1` resource needs to be replaced.

**Comment by outcatcher** [comment]
Additional problem: update of `certificate_id` is ignored.

**Comment by outcatcher** [comment]
@DasLeo 
I don't see really good way fixing this on the provider side, but `create_before_destroy` + random name approach will work in the release `v1.25.1` (certificate update fixed in #1294), e.g.:

```hcl
locals {
  cert = sensitive(file("./certificate.crt"))
  key = sensitive(file("./private.key"))
}

resource "random_integer" "postfix" {
  min = 0
  max = 100000

  keepers = {
    cert = local.cert
    key  = local.key
  }
}

resource "opentelekomcloud_waf_certificate_v1" "cert" {
  name    = "test-cert-${random_integer.postfix.result}"
  content = local.cert
  key     = local.key

  lifecycle {
    create_before_destroy = true
  }
}

resource "opentelekomcloud_networking_floatingip_v2" "eip" {}

resource "opentelekomcloud_waf_domain_v1" "domain" {
  hostname       = "acme.com"
  server {
    address         = opentelekomcloud_networking_floatingip_v2.eip.address
    client_protocol = "HTTPS"
    server_protocol = "HTTPS"
    port            = "443"
  }
  certificate_id = opentelekomcloud_waf_certificate_v1.cert.id
  proxy          = false
  cipher         = "cipher_2"
  tls            = "TLS v1.2"
}

```
